### PR TITLE
Optimize blobstore gc when current stat total size is 0

### LINKED ISSUE #5016: Blob may not cleanup after restore.
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
- write some page with a long snapshot
- Then restore, total size may become 0.
middle_img_v2_6a3db914-5c11-4b81-b2ee-c7c8f837aa4g

### PR DESCRIPTION
**Pull request by jiaqizho**
### What problem does this PR solve?

Issue Number: close #5016

Problem Summary:
- After `restore`,  the size of `snapshot` will be 0, Then some of the total size from blobstat will get 0.
- But the blob file size in disk may not be 0.

### What is changed and how it works?
- Truncate the blob file size.
- Also update valid rate to 0 is total size is 0.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/Page/V3/BlobStore.cpp [hunk 1]
```diff
@@ -914,7 +914,14 @@ std::vector<BlobFileId> BlobStore::getGCStats()
...
             if (right_margin == 0)
             {
-                LOG_FMT_TRACE(log, "Current blob is empty [blob_id={}, total size(all invalid)={}].", stat->id, stat->sm_total_size);
+                assert(stat->sm_valid_rate == 0);
+                LOG_FMT_TRACE(log, "Current blob is empty [blob_id={}, total size(all invalid)={}] [valid_rate={}].", stat->id, stat->sm_total_size, stat->sm_valid_rate);
+                // If current blob empty, the size of in disk blob may not empty
+                // So we need truncate current blob, and let it be reused.
+                auto blobfile = getBlobFile(stat->id);
+                LOG_FMT_TRACE(log, "Truncate empty blob file [blob_id={}] to 0.", stat->id, stat->sm_total_size, right_margin);
+                blobfile->truncate(0);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/BlobStore.cpp:923]:
```suggestion
                blobfile->truncate(right_margin);
```

File: dbms/src/Storages/Page/V3/BlobStore.cpp [hunk 2]
```diff
@@ -914,7 +914,14 @@ std::vector<BlobFileId> BlobStore::getGCStats()
             // Avoid divide by zero
             if (right_margin == 0)
             {
-                LOG_FMT_TRACE(log, "Current blob is empty [blob_id={}, total size(all invalid)={}].", stat->id, stat->sm_total_size);
+                assert(stat->sm_valid_rate == 0);
+                LOG_FMT_TRACE(log, "Current blob is empty [blob_id={}, total size(all invalid)={}] [valid_rate={}].", stat->id, stat->sm_total_size, stat->sm_valid_rate);
+                // If current blob empty, the size of in disk blob may not empty
+                // So we need truncate current blob, and let it be reused.
+                auto blobfile = getBlobFile(stat->id);
+                LOG_FMT_TRACE(log, "Truncate empty blob file [blob_id={}] to 0.", stat->id, stat->sm_total_size, right_margin);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/BlobStore.cpp:922]:
```suggestion
                LOG_FMT_TRACE(log, "Truncate empty blob file [blob_id={}] to 0.", stat->id);
```

File: dbms/src/Storages/Page/V3/BlobStore.h [hunk 3]
```diff
@@ -107,7 +107,7 @@ class BlobStore : private Allocator<false>
             UInt64 sm_max_caps = 0;
             UInt64 sm_total_size = 0;
             UInt64 sm_valid_size = 0;
-            double sm_valid_rate = 1.0;
+            double sm_valid_rate = 0;
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/BlobStore.h:110]:
```suggestion
            double sm_valid_rate = 0.0;
```

File: dbms/src/Storages/Page/V3/BlobStore.cpp [hunk 4]
```diff
@@ -1468,7 +1475,7 @@ void BlobStore::BlobStats::BlobStat::recalculateSpaceMap()
     const auto & [total_size, valid_size] = smap->getSizes();
     sm_total_size = total_size;
     sm_valid_size = valid_size;
-    sm_valid_rate = valid_size * 1.0 / total_size;
+    sm_valid_rate = total_size == 0 ? 0 : valid_size * 1.0 / total_size;
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/BlobStore.cpp:1478]:
```suggestion
    sm_valid_rate = total_size == 0 ? 0.0 : valid_size * 1.0 / total_size;
```

**Review by JaySon-Huang — APPROVED** [review]
LGTM with small comments

Use "0.0" indicate it is double type
Take care of the param of LOG_FMT

**Review by CalvinNeo — APPROVED** [review]
lgtm

File: dbms/src/Storages/Page/V3/BlobStore.cpp [hunk 5]
```diff
@@ -914,7 +914,14 @@ std::vector<BlobFileId> BlobStore::getGCStats()
             // Avoid divide by zero
             if (right_margin == 0)
             {
-                LOG_FMT_TRACE(log, "Current blob is empty [blob_id={}, total size(all invalid)={}].", stat->id, stat->sm_total_size);
+                assert(stat->sm_valid_rate == 0);
```
Comment (flowbehappy) [inline dbms/src/Storages/Page/V3/BlobStore.cpp:917]: Please don't use assert outside UT code. You can use `RUNTIME_ASSERT` or simply throw expcetions.

File: dbms/src/Storages/Page/V3/BlobStore.cpp [hunk 6]
```diff
@@ -914,7 +914,22 @@ std::vector<BlobFileId> BlobStore::getGCStats()
             // Avoid divide by zero
             if (right_margin == 0)
             {
-                LOG_FMT_TRACE(log, "Current blob is empty [blob_id={}, total size(all invalid)={}].", stat->id, stat->sm_total_size);
+                if (stat->sm_valid_rate == 0)
```
Comment (flowbehappy) [inline dbms/src/Storages/Page/V3/BlobStore.cpp:917]:
```suggestion
                if (unlikely(stat->sm_valid_rate == 0))
```

File: dbms/src/Storages/Page/V3/BlobStore.cpp [hunk 7]
```diff
@@ -914,7 +914,22 @@ std::vector<BlobFileId> BlobStore::getGCStats()
             // Avoid divide by zero
             if (right_margin == 0)
             {
-                LOG_FMT_TRACE(log, "Current blob is empty [blob_id={}, total size(all invalid)={}].", stat->id, stat->sm_total_size);
+                if (unlikely(stat->sm_valid_rate == 0))
```
Comment (flowbehappy) [inline dbms/src/Storages/Page/V3/BlobStore.cpp:917]:
```suggestion
                if (unlikely(stat->sm_valid_rate != 0))
```

File: dbms/src/Storages/Page/V3/BlobStore.cpp [hunk 8]
```diff
@@ -914,7 +914,22 @@ std::vector<BlobFileId> BlobStore::getGCStats()
...
+                                                stat->sm_valid_rate));
+                }
+
+                LOG_FMT_TRACE(log, "Current blob is empty [blob_id={}, total size(all invalid)={}] [valid_rate={}].", stat->id, stat->sm_total_size, stat->sm_valid_rate);
+
+                // If current blob empty, the size of in disk blob may not empty
+                // So we need truncate current blob, and let it be reused.
+                auto blobfile = getBlobFile(stat->id);
+                LOG_FMT_TRACE(log, "Truncate empty blob file [blob_id={}] to 0.", stat->id);
+                blobfile->truncate(right_margin);
```
Comment (flowbehappy) [inline dbms/src/Storages/Page/V3/BlobStore.cpp:931]: It is safe to do truncate without lock protection?

Comment (flowbehappy) [reply dbms/src/Storages/Page/V3/BlobStore.cpp:931]: OK. Just see `lock` before.

**Pull request merged**